langpuppet/merge.go

```diff
@@ -0,0 +1,173 @@
+// Mgmt
+// Copyright (C) 2013-2018+ James Shubin and the project contributors
+// Written by James Shubin <james@shubin.ca> and the project contributors
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+// Package langpuppet implements an integration entrypoint that combines lang and Puppet.
+package langpuppet
+
+import (
+	"fmt"
+	"github.com/purpleidea/mgmt/engine"
+	"github.com/purpleidea/mgmt/pgraph"
+	"strings"
+)
+
+// This method returns the merged graph containing all vertices
+// and edges found in the graphs produced by the lang and Puppet GAPIs
+// associated with the wrapping GAPI.
+// Vertices are merged if they adhere to the following rules (for any
+// given value of POSTFIX):
+//  1. The graph from lang contains a noop vertex named puppet_POSTFIX
+//  2. The graph from Puppet contains an empty class mgmt_POSTFIX
+//  3. The resulting graph will contain one noop vertex named POSTFIX
+//     that replaces all nodes mentioned in (1) and (2)
+// All edges connecting to any of the vertices merged this way
+// will be present in the merged graph.
+func (obj *GAPI) mergeChildGraphs() (*pgraph.Graph, error) {
+	lg := obj.currentLangGraph
+	pg := obj.currentPuppetGraph
+
+	if lg == nil || pg == nil {
+		return nil, fmt.Errorf("cannot merge graphs until both child graphs are loaded")
+	}
+
+	result, err := pgraph.NewGraph(lg.Name + "+" + pg.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	merge_targets := make(map[string]pgraph.Vertex)
+
+	// first add all vertices from the lang graph
+	for _, vertex := range lg.Vertices() {
+		if strings.Index(vertex.String(), "noop[puppet_") == 0 {
+			resource, ok := vertex.(engine.NamedRes)
```

Comment:
This looks suspicious... Maybe it's okay though. Is it what you want or really need? Why not just `Res` ?

Reply:
For the vertex merging, I need to inspect the `Name` of all `noop` resources.

Reply:
...which I now realize would work with `Res` just as well, so yes, will probably do it.

Reply:
(unless you end up making the `named` aspect optional, as a code comment suggests ;)

Reply:
Anyway, fixed.